File: update_files/Update_players_table.py
import pymysql
import csv
import datetime

def update_players_info(file_path):
    try:
        # a cursor is an object that allows me to retrieve and manipulate data from a result set
        # the cursor is what allows me to run the SQL query to gather all of the data
        # cursors usually return the results as tuples, but different cursor return the data in different ways


        # connects to my MySQL DB and allows me to query from it
        # this is a very similar set up to the R version of doing this
        connection = pymysql.connect(host='****',
                                    user='*********',
                                    password='*******',
                                    database='******',
                                    # this code tells pymysql to choose the cursor that returns the reuslts as dictionary
                                    # this gives me column names as the keys and the column values as the values
                                    cursorclass=pymysql.cursors.DictCursor
                                    )

        # create the cursor object
        cursor = connection.cursor()

        # create a temp table to hold our reference data
        # this has the same layout as my players table
        create_table_query = """
        create temporary table current_players (
            player_id int NOT NULL,
            player_name varchar(100),
            position_abbrv varchar(5),
            position_type varchar(15),
            status_code varchar(5),
            status_description varchar(50),
            team_id int NOT NULL,
            statcast_pos_lookup varchar(10)
        );
        """

        # execute the query and print an update statement
        cursor.execute(create_table_query)
        print("Temp Table Created...")

        # the csv file is generated from the 'update_players_table.r' file
        # the way the table is generated in R, it has no header so we dont need to skip a line
        print("Inserting Data...")
        with open(file_path, "r") as current_player_info:
            csv_reader = csv.reader(current_player_info)
            for row in csv_reader:
                # this is the query we want to execute in SQL to add data to the table
                # the %s acts a placeholder, when we execute the query, Python inserts the actual values of the row
                insert_query = """
                INSERT INTO current_players (player_id, player_name, position_abbrv, position_type, status_code, status_description, team_id, statcast_pos_lookup)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
                """
                cursor.execute(insert_query, row)
        
        print("Data Successfully Inserted!")

        # this where we actually update the players table 
        # its safer to update everything since there could be multiple things that change at once
        print("Updating Table...")
        update_query = """
        update players
        join current_players on players.player_id = current_players.player_id
        set players.player_id = current_players.player_id,
            players.player_name = current_players.player_name,
            players.position_abbrv = current_players.position_abbrv,
            players.position_type = current_players.position_type,
            players.status_code = current_players.status_code,
            players.status_description = current_players.status_description,
            players.team_id = current_players.team_id,
            players.statcast_pos_lookup = current_players.statcast_pos_lookup;
        """

        cursor.execute(update_query)

        # we could run into a situation where a new player joins the league
        # if that is the case, we want to insert them into our table
        insert_new_players_query = """
        insert into players (player_id, player_name, position_abbrv, position_type, status_code, status_description, team_id, statcast_pos_lookup)
        select * 
        from current_players
        where player_id not in (select player_id from players);
        """

        cursor.execute(insert_new_players_query)
        
        # when we are finished, we want to drop the temp table from our database
        drop_table_query = """
        drop temporary table if exists current_players;
        """

        cursor.execute(drop_table_query)

        print("Table Updated and Temp Table Removed")

        print("Updating Transactions table...")
        with open(file_path, "r") as current_player_info:
            csv_reader = csv.reader(current_player_info)
            for row in csv_reader:
                row.append(datetime.date.today().strftime("%Y-%m-%d"))
                # this is the query we want to execute in SQL to add data to the table
                # the %s acts a placeholder, when we execute the query, Python inserts the actual values of the row
                insert_query = """
                INSERT INTO transactions (player_id, player_name, position_abbrv, position_type, status_code, status_description, team_id, statcast_pos_lookup, transaction_date)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
                """
                cursor.execute(insert_query, row)

        print("Transactions table updated!")

        # close the cursor and the connection when finished
        cursor.close() 
        connection.commit() 
        connection.close()

        print("Update Completed!")
    except:
        cursor.close()
        connection.close()
        raise



update_players_info("/Users/drewc/OneDrive/Documents/2024/MLB Project 2024/update_files/mlb2024_updated_players.csv")

File: update_files/update_players_table.r
# ^ UPDATING THE PLAYERS TABLE

# ^ WHEN TO RUN: probably everyday. guys are being sent down and called up every day. trades, free agent signings, injuries, etc
# ^ SCRIPT GOAL: find all of the mismatched records in my database and the most updated version from the baseballr package
# ^ INPUTS: no direct inputs from the user, we get all records from my database and the most updated table from the package
# ^ OUTPUT: write a csv file containing all of the records that did not match
# ^ PACKAGE: baseballr is doing most of the heavy lifting in terms of collecting the data (https://rdrr.io/github/BillPetti/baseballr/man/baseballr-package.html)
# ^ TO RUN: press [CTRL + A] and then [CTRL + Enter]


library(DBI)
library(RMySQL)
library(baseballr)
library(dplyr)
library(tidyr)

# ? #####################################################################################
# ? GET THE CURRENT VERSION OF MY PLAYERS TABLE
# ? #####################################################################################

# connect to the database
# send my query and fetch the results into the 'my_players' variable
# remove the variables used to get the data and then disconnect from the database
con <- dbConnect(MySQL(),
                user = "******",
                password = "*******",
                host = "*****",
                dbname = "******")


query <- "select * from players;"

result <- dbSendQuery(con, query)

# n = -1 means to return all rows, by default there is a limit to how many rows are returned
my_players <- fetch(result, n = -1)
dbClearResult(result)
rm(result)
rm(query)
dbDisconnect(con)
rm(con)


# ? #####################################################################################
# ? GET THE UPDATED VERSION OF THE PLAYERS TABLE
# ? #####################################################################################

# in order to get the rosters for each team, we need to get each of the team IDs
# doing it this way means that if any team ID changes for some reason, our code wont break
teams <- mlb_teams(season = 2024, sport_ids = c(1)) %>%
  select(team_id)

# create our own function that takes the team ID and returns the 40 man roster for that team in the 2024 season
# we also trim down the columns of this and only choose relevant columns
get_mlb_rosters <- function(x){
  
  trimmed_player_info_df <- mlb_rosters(team_id = x, season = 2024, roster_type = "fullRoster") %>%
    select(person_id, 
      person_full_name, 
      position_abbreviation, 
      position_type, 
      status_code, 
      status_description, 
      team_id)
  
  return(trimmed_player_info_df)
}

# we then use the function that was created above to collect player data
# use lapply to go through each team ID from the teams df we got above 
# this returns a list of data frames, then we can use rbind to combine them all into one dataframe
players_list <- lapply(X = teams$team_id, 
                       FUN = function(x){get_mlb_rosters(x)})

curr_players <- do.call("rbind", players_list)

# add this column that will allows us to properly look up a players stats based on what position they play
curr_players$statcast_pos_lookup <- ifelse(curr_players$position_type == "Pitcher", "pitcher", "batter")

# remove any duplicated players
# still unsure why this was happening...
curr_players <- curr_players[!duplicated(curr_players$person_id), ]

# remove "." or other special characters from the names, keep the "-" (dashes)
# iconv removes accent marks in our player names, keeps it consistent
curr_players$person_full_name <- iconv(curr_players$person_full_name, to = "ASCII//TRANSLIT")
curr_players$person_full_name <- gsub("\\.", "", curr_players$person_full_name)

# change the columns to match the ones found in our database
colnames(curr_players) <- c("player_id", "player_name", "position_abbrv", "position_type", "status_code", "status_description", "team_id", "statcast_pos_lookup")


# ? #####################################################################################
# ? FIND THE MISMATCHED ROWS AND WRITE THEM TO A FILE
# ? #####################################################################################

# this returns all of the players in the most updated list from mlb that do not have a match in my list
# this means that these players had some sort of change to them and I need to update my table
# I chose these three columns to look for because we can add a new player, change teams, or change status (injury, called up, etc)
mismatched_players <- anti_join(curr_players, my_players, by = c("player_id", "team_id", "status_code"))

# write this data to a csv file to be read and processed through Python
write.table(mismatched_players, "/Users/drewc/OneDrive/Documents/2024/MLB Project 2024/update_files/mlb2024_updated_players.csv", row.names = FALSE, col.names = FALSE, sep = ",",quote = FALSE)
